### ISSUE TITLE
Using @Type with v 0.0.6 throws a crypto not found error

### ISSUE THREAD

**Issue by the-ult** [pleerock/constructor-utils#3]
After installing serializer.ts v0.0.6 and using the @Type decorator I keep getting the following error:
`localhost/:99 Error: Error: XHR error (404 Not Found) loading http://localhost:5555/node_modules/crypto/package.json(…)`

```
import {Type} from 'serializer.ts/Decorators';

export class Invoice {
  @Type(() => Number)
  id: number;
}
```

**Comment by pleerock** [comment]
do you get this error after update? Or its not only related to 0.0.6 version issue? `crypto` is used by `reflect-metadata` package, which is required when you are using decorators in typescript. Maybe you'll try to install it?

**Comment by the-ult** [comment]
I tried it both with reflect-meta 0.1.2 and 0.1.3,
(0.1.2 is needed by Angular2.)
Could try it again with 0.0.4 to see if it does work with that version.

I got the error as soon as I imported the serializer.ts/Decorators.

Op vr 18 mrt. 2016 20:58 schreef Umed Khudoiberdiev <
notifications@github.com>:

> do you get this error after update? Or its not only related to 0.0.6
> version issue? crypto is used by reflect-metadata package, which is
> required when you are using decorators in typescript. Maybe you'll try to
> install it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/pleerock/serializer.ts/issues/3#issuecomment-198518626>
>
-- 
met vriendelijke groet,
Arjen

M 06-51 800 419
W www.payt.nl

**Comment by pleerock** [comment]
Personally I use this library on both frontend and backend. On the frontend looks like I installed `crypto` and have it in my systemjs dependency configuration. On the backend I didn't do it since crypto is already shipped with node.js. crypto is required by reflect-metadata package [here](https://github.com/rbuckton/ReflectDecorators/blob/master/Reflect.ts#L1398) but in their code there is a `isNode` check, not sure why it triggers on the frontend. It maybe a reflect-metadata issue, or maybe something related to the way systemjs works. [Here](https://github.com/jspm/npm/issues/77) is something on systemjs repo that can help. 

For now I'll suggest either to install crypto, either use empty configuration like this: `{ 'map': { 'crypto': '@empty' } }`. I'll need to experiment with it and go deeper to understand what is going on and is there a possibility for it to work without crypto. If you'll get some additional info, please let me know.

**Comment by the-ult** [comment]
Thanks for your reply and the samples.
Which crypto have you installed? (I'm using npm) and where do I put the `{ 'map': { 'crypto': '@empty' } }` tsconfig.json does't have a map option/setting. Can't seem to find it.

**Comment by pleerock** [comment]
What module bundler are you using? Im using system.js (as in the angular2 tutorials and samples). You can use `System.config({ map: { ... } });` to config it. [docs](https://github.com/systemjs/systemjs#browser)

**Comment by the-ult** [comment]
Found it in the System.config indeed. Helped, but got another error. 
Got some more time tomorrow. Then I'll dive into it some more and try to give some better feedback/logs

**Comment by the-ult** [comment]
(updated to serializer.ts 0.0.10, angular2 0.1.2)

After using the setting in System.config I keep getting:
`Cannot read property 'getOptional' of undefined`

Can't find what causes this error.

I'm afraid the Typing decorator does not seem to work.
Do you happen to have an example of a frontend project which works?

**Comment by pleerock** [comment]
@the-ult did 0.0.9 worked for you? if yes then maybe the easiest for you to keep going with it. If you are good to changes, then try to update to 0.0.11 . if does not help and you are still curious try to debug, or provide more information or even a project example, I'll debug instead of you. There is no `getOptional` method in the library, that's why I cannot tell you anything.

Unfortunately I dont have a sample frontend project that I can publish, only private one. I'll try to create a sample project later when I'll have more time. I tested 0.0.11 works same as previous versions on me.

**Comment by pleerock** [comment]
These are two gist: [config.js](https://gist.github.com/pleerock/28cc0b7b012d8b61d1f1) and [index.html](https://gist.github.com/pleerock/70c245c4fcdf57cef502) from a private project. Maybe this can help you

**Comment by pleerock** [comment]
ah and yes I'm using jspm, but I remember I used serializer without it before too, but configuration was different

**Comment by the-ult** [comment]
Hey @pleerock, thanks for your help, but I still can't get the @Type to work.

Tried to install `crypto-browserify` which is used bij `"github:jspm/nodelibs-crypto@0.1.0"` and define it as `crypto` in my system.config. (I'm using [angular2-seed](https://github.com/mgechev/angular2-seed) as base-project)
I keep getting one of these messages:

* `XHR error (404 Not Found) loading http://localhost:5555/node_modules/crypto/package.json(…)`
* missing child process
  - after setting/using `crypto-browserify`
* `Cannot read property 'getOptional' of undefined`
  - after setting: `{ 'map': { 'crypto': '@empty' } }`

The serializing itself works, but just the @Type decorator and @Skip just don't seem to work. (What is crypto used for in Decorator.ts?)
I'm gonna leave it as is and just use the (de)serializing for now. Have to proceed with the rest of the project.

Gonna try again/spent some time on it in a few weeks.

Thanks again for your help.

**Comment by pleerock** [comment]
instead of making it `'crypto': '@empty'` maybe you simply try to install `crypto` (and map it if needed in config.js)? type decorator and skip decorator are not using crypto package, but third party dependency - reflect-metadata uses.

**Comment by the-ult** [comment]
But which crypto?
I tried crypto, crypto-browserify and crypto-js.
And the npm install crypto, installs a really old package, which can't be
the correct one.

I haven't got any problems with reflect-meta and decorators with other
libraries, so that's the strange part.

I could ask somebody from angular2-seed of they have any idea..

Op wo 30 mrt. 2016 18:03 schreef Umed Khudoiberdiev <
notifications@github.com>:

> instead of making it 'crypto': '@empty' maybe you simply try to install
> crypto (and map it if needed in config.js)? type decorator and skip
> decorator are not using crypto package, but third party dependency -
> reflect-metadata uses.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/pleerock/serializer.ts/issues/3#issuecomment-203504376>
>
-- 
met vriendelijke groet,
Arjen

M 06-51 800 419
W www.payt.nl

**Comment by the-ult** [comment]
Hey @pleerock,

I think I found the cause: https://github.com/pleerock/constructor-utils/blob/master/src/decorators.ts
`import "reflect-metadata"; 
When I remove the import/require reflect-metadata it works perfectly.

See: https://github.com/angular/angular/issues/5958
and:
https://github.com/angular/angular/blob/master/modules/angular2/docs/bundles/overview.md#required-polyfills
> In short: don't load zones nor reflect-metadata through system loader. Include angular2-polyfills.js using a script tag instead. 

https://github.com/angular/angular/issues/5306
> Reflect-metadata is a shim and as any other shim you shouldn't try to import it but rather make sure that the corresponding file is included in a browser for a test runner (via script tag or otherwise).
> This is the same mechanism as for any other shim - you just need to load them as soon as possible not import them.

**Comment by pleerock** [comment]
hey @the-ult nice found. Didn't know that reflect-metadata is a shim. Yes, probably you are right - I need to remove import from decorator.ts and release a new version. Going to do it in few next days.

**Comment by the-ult** [comment]
Great thanks! Really appreciated!
If I find any other things to contribute I'll let you know.

Op ma 2 mei 2016 22:53 schreef Umed Khudoiberdiev <notifications@github.com
>:

> hey @the-ult <https://github.com/the-ult> nice found. Didn't know that
> reflect-metadata is a shim. Yes, probably you are right - I need to remove
> import from decorator.ts and release a new version. Going to do it in few
> next days.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/pleerock/constructor-utils/issues/3#issuecomment-216359933>
>
-- 
met vriendelijke groet,
Arjen

M 06-51 800 419
W www.payt.nl

**Comment by pleerock** [comment]
Issue has been fixed, thank you for contribution and research

**Comment by the-ult** [comment]
Thanks for fixing it!

Op zo 8 mei 2016 21:53 schreef Umed Khudoiberdiev <notifications@github.com
>:

> Closed #3 <https://github.com/pleerock/constructor-utils/issues/3>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/pleerock/constructor-utils/issues/3#event-653789580>
>
-- 
met vriendelijke groet,
Arjen

M 06-51 800 419
W www.payt.nl
